[Doc-Improvement] Add CDC-Ingestion section (#1697)
NetEase/arctic: docs/user-guides/cdc-ingestion.md

        parent: User Guides
        weight: 400
---
# CDC Ingestion

> CDC stands for Change Data Capture, which is a broad concept, as long as it can capture the change data, it can be called CDC. [Flink CDC](https://github.com/ververica/flink-cdc-connectors) is a Log message-based data capture tool, all the inventory and incremental data can be captured. Taking MySQL as an example, it can easily capture Binlog data through Debezium and process the calculations in real time to send them to the Arctic data lake. The Arctic data lake can then be queried by other engines.

Reviewer: Arctic's positioning is Lakehouse management service, `Arctic data lake` is an inaccurate statement.I suggest changing it directly to "data lake".

> This section will show how to practice one table into the lake and multiple tables into the lake for both [Native-Iceberg](../formats/iceberg.md) and [Mixed-Iceberg](../formats/mixed-iceberg.md) format.

Reviewer: ```suggestion
This section will show how to ingest one table or multiple tables into the data lake for both [Iceberg](../formats/iceberg.md) format and [Mixed-Iceberg](../formats/mixed-iceberg.md) format.
```

> ## One table into the lake

Reviewer: ```suggestion
## Ingest into one table
```

> ### Native Iceberg format

Reviewer: ```suggestion
### Iceberg format
```

> The following example will show how MySQL CDC data is written to a Native-Iceberg table.

Reviewer: ```suggestion
The following example will show how MySQL CDC data is written to a Iceberg table.
```

```sql
CREATE TABLE products (
    id INT,
    name STRING,
    description STRING,
    PRIMARY KEY (id) NOT ENFORCED
) WITH (
    'connector' = 'mysql-cdc',
    'hostname' = 'localhost',
    'port' = '3306',
    'username' = 'root',
    'password' = '123456',
    'database-name' = 'mydb',
    'table-name' = 'products'
);
  
CREATE CATALOG iceberg_hadoop_catalog WITH (
    'type'='iceberg',
    'catalog-type'='hadoop',
    'warehouse'='hdfs://nn:8020/warehouse/path',
    'property-version'='1'
);

CREATE TABLE IF NOT EXISTS `iceberg_hadoop_catalog`.`default`.`sample` (
    id INT,
    name STRING,
    description STRING,
    PRIMARY KEY (id) NOT ENFORCED
);

INSERT INTO `iceberg_hadoop_catalog`.`default`.`sample` SELECT * FROM products;
```

### Mixed Iceberg format
The following example will show how MySQL CDC data is written to a Mixed-Iceberg table.

**Requirements**

> Please add [Flink Connector MySQL CDC](https://repo1.maven.org/maven2/com/ververica/flink-connector-mysql-cdc/2.3.0/flink-connector-mysql-cdc-2.3.0.jar) and [Arctic](https://repo1.maven.org/maven2/com/netease/arctic/arctic-flink-runtime-1.14/0.4.1/arctic-flink-runtime-1.14-0.4.1.jar) Jars to the lib directory of the Flink engine package.

Reviewer: ```suggestion
Please add [Flink Connector MySQL CDC](https://repo1.maven.org/maven2/com/ververica/flink-connector-mysql-cdc/2.3.0/flink-connector-mysql-cdc-2.3.0.jar) and [Arctic](../../download) Jars to the lib directory of the Flink engine package.
```

```sql
CREATE TABLE products (
    id INT,
    name STRING,
    description STRING,
    PRIMARY KEY (id) NOT ENFORCED
) WITH (
    'connector' = 'mysql-cdc',
    'hostname' = 'localhost',
    'port' = '3306',
    'username' = 'root',
    'password' = '123456',
    'database-name' = 'mydb',
    'table-name' = 'products'
);

CREATE CATALOG arctic_catalog WITH (
    'type'='arctic',
    'metastore.url'='thrift://<ip>:<port>/<catalog_name_in_metastore>'
); 

CREATE TABLE IF NOT EXISTS `arctic_catalog`.`db`.`test_tb`(
    id INT,
    name STRING,
    description STRING,
    PRIMARY KEY (id) NOT ENFORCED
);

INSERT INTO `arctic_catalog`.`db`.`test_tb` SELECT * FROM products;
```

> ## Multiple tables into the lake

Reviewer: ```suggestion
## Ingest Into multiple tables
```

> ### Native Iceberg format

Reviewer: ```suggestion
### Iceberg format
```

The following example will show how to write CDC data from multiple MySQL tables into the corresponding Native-Iceberg table.

**Requirements**

Please add [Flink Connector MySQL CDC](https://mvnrepository.com/artifact/com.ververica/flink-connector-mysql-cdc/2.3.0) and [Iceberg](https://mvnrepository.com/artifact/org.apache.iceberg/iceberg-flink-1.14/1.1.0) dependencies to your Maven project's pom.xml file.

> ```java

Reviewer: add import statements at first

        .append();
    }

    env.execute("Sync MySQL to Native Iceberg table");
  }

  static class CompositeDebeziumDeserializationSchema
    implements DebeziumDeserializationSchema<RowData> {

    private final Map<String, RowDataDebeziumDeserializeSchema> deserializationSchemaMap;

    public CompositeDebeziumDeserializationSchema(
      final Map<String, RowDataDebeziumDeserializeSchema> deserializationSchemaMap) {
      this.deserializationSchemaMap = deserializationSchemaMap;
    }

    @Override
    public void deserialize(final SourceRecord record, final Collector<RowData> out)
      throws Exception {
      final Struct value = (Struct) record.value();
      final Struct source = value.getStruct("source");
      final String db = source.getString("db");
      final String table = source.getString("table");
      if (deserializationSchemaMap == null) {
        throw new IllegalStateException("deserializationSchemaMap can not be null!");
      }
      deserializationSchemaMap.get(db + "." + table).deserialize(record, out);
    }

    @Override
    public TypeInformation<RowData> getProducedType() {
      return TypeInformation.of(RowData.class);
    }
  }

>   static class RowDataVoidProcessFunction extends ProcessFunction<RowData, Void> {

Reviewer: Give a more specific name to the class.

    id INT,
    name STRING,
    description STRING,
    PRIMARY KEY (id) NOT ENFORCED
);

insert into `iceberg_hadoop_catalog`.`default`.`sample` select * from products;
```

### Mixed Iceberg format
The following example will show how MySQL CDC data is written to a Mixed-Iceberg table.
```sql
CREATE TABLE products (
    id INT,
    name STRING,
    description STRING,
    PRIMARY KEY (id) NOT ENFORCED
) WITH (
    'connector' = 'mysql-cdc',
    'hostname' = 'localhost',
    'port' = '3306',
    'username' = 'root',
    'password' = '123456',
    'database-name' = 'mydb',
    'table-name' = 'products'
);

CREATE CATALOG arctic_catalog WITH (
    'type'='arctic',
    'metastore.url'='thrift://<ip>:<port>/<catalog_name_in_metastore>'
); 

CREATE TABLE IF NOT EXISTS `arctic_catalog`.`db`.`test_tb`(
    id INT,
    name STRING,
    description STRING,
    PRIMARY KEY (id) NOT ENFORCED
);

insert into `arctic_catalog`.`db`.`test_tb` select * from products;
```

## Multiple tables into the lake
### Native Iceberg format
The following example will show how to write CDC data from multiple MySQL tables into the corresponding Native-Iceberg table.
```java

> import com.ververica.cdc.connectors.mysql.source.MySqlSource;

Reviewer 2: I think we can remove the import, and add maven or gradle dependency detail at the head of demos

    }

    @Override
    public void deserialize(final SourceRecord record, final Collector<RowData> out)
      throws Exception {
      final Struct value = (Struct) record.value();
      final Struct source = value.getStruct("source");
      final String db = source.getString("db");
      final String table = source.getString("table");
      if (deserializationSchemaMap == null) {
        throw new IllegalStateException("deserializationSchemaMap can not be null!");
      }
      deserializationSchemaMap.get(db + "." + table).deserialize(record, out);
    }

    @Override
    public TypeInformation<RowData> getProducedType() {
      return TypeInformation.of(RowData.class);
    }
  }

  static class RowDataVoidProcessFunction extends ProcessFunction<RowData, Void> {
    private final Map<String, RowRowConverter> converters;

    public RowDataVoidProcessFunction(final Map<String, RowRowConverter> converterMap) {
      this.converters = converterMap;
    }

    @Override
    public void processElement(final RowData rowData,
                               final ProcessFunction<RowData, Void>.Context ctx, final Collector<Void> out)
      throws Exception {

>       final String key = rowData.getString(rowData.getArity() - 2).toString();

Reviewer 2: why minus 2?

      .hostname("yourHostname")
      .port(yourPort)
      .databaseList("test_db")
      // setting up tables to be captured
      .tableList("test_db.user", "test_db.product")
      .username("yourUsername")
      .password("yourPassword")
      .deserializer(new CompositeDebeziumDeserializationSchema(debeziumDeserializeSchemas))
      .build();

    StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

    // enable checkpoint
    env.enableCheckpointing(60000);

    // Split CDC streams by table name
    SingleOutputStreamOperator<Void> process = env
      .fromSource(mySqlSource, WatermarkStrategy.noWatermarks(), "MySQL Source").setParallelism(4)
      .process(new RowDataVoidProcessFunction(pathAndTable.stream()
        .collect(toMap(e -> e.f0.toString(),
          e -> RowRowConverter.create(e.f1.getResolvedSchema().toPhysicalRowDataType())))))
      .name("split stream");

    // create Arctic sink and insert into cdc data
    InternalCatalogBuilder catalogBuilder = InternalCatalogBuilder.builder().metastoreUrl(
      "thrift://<ip>:<port>/<catalog_name_in_metastore>");

>     Map<String, TableSchema> tableSchemas = new HashMap<>();

Reviewer 2: How about renaming to sinkTableSchemas?

    }

    @Override
    public void deserialize(final SourceRecord record, final Collector<RowData> out)
      throws Exception {
      final Struct value = (Struct) record.value();
      final Struct source = value.getStruct("source");
      final String db = source.getString("db");
      final String table = source.getString("table");
      if (deserializationSchemaMap == null) {
        throw new IllegalStateException("deserializationSchemaMap can not be null!");
      }
      deserializationSchemaMap.get(db + "." + table).deserialize(record, out);
    }

    @Override
    public TypeInformation<RowData> getProducedType() {
      return TypeInformation.of(RowData.class);
    }
  }

  static class RowDataVoidProcessFunction extends ProcessFunction<RowData, Void> {
    private final Map<String, RowRowConverter> converters;

    public RowDataVoidProcessFunction(final Map<String, RowRowConverter> converterMap) {
      this.converters = converterMap;
    }

    @Override
    public void processElement(final RowData rowData,
                               final ProcessFunction<RowData, Void>.Context ctx, final Collector<Void> out)
      throws Exception {

>       final String key = rowData.getString(rowData.getArity() - 2).toString();

Reviewer 2: why minus 2?
